src/c-api.adoc

```diff
@@ -814,9 +814,114 @@ statements, including both RISC-V specific and common operand modifiers.
 [id=function-multi-version]
 == Function Multi-version
 
-Function multi-versioning(FMV) provides an approach to selecting the appropriate
-function according to the runtime environment. The final binary may contain all
-versions of the function, with the compiler generating all supported versions
-and the runtime selecting the appropriate one.
+Function multi-versioning (FMV) allows selecting the appropriate function based on the runtime environment. The binary may contain multiple versions of the function, with the compiler generating all supported versions and selecting the appropriate one at runtime.
 
-This feature is triggered by `target_version/target_clones` function attribute.
+This feature is activated by the `target_version/target_clones` function attributes.
+
+=== Extension Bitmask
+
+The Extension Bitmask is used to probe whether features are enabled during runtime. This is achieved through three bitmask structures: `__riscv_feature_bits` for standard extensions,  `__riscv_vendor_feature_bits` for vendor-specific extensions and `__riscv_cpu_model` for the CPU model. Additionally, `__init_riscv_feature_bits` is used to update the contents of these structures according to the system configuration.
+
+The bitmask structures use the following definitions:
+
+```c
+struct {
+    unsigned length;
+    unsigned long long features[];
+} __riscv_feature_bits;
+
+struct {
+    unsigned length;
+    unsigned long long features[];
+} __riscv_vendor_feature_bits;
+
+struct {
+  unsigned mvendorid;
+  unsigned marchid;
+  unsigned mimpid;
+} __riscv_cpu_model;
+```
+
+- `length`: Represents the number of elements in the features array.
+- `features`: An `unsigned long long` array where each bit indicates 1 for a specific extension enabled by the system or 0 for the extension's status unknown in system.
+- `mvendorid`: Indicates the value of `mvendorid` CSR.
+- `marchid`: Indicates the value of `marchid` CSR.
+- `mimpid`: Indicates the value of `mimpid` CSR.
+
+To initiate these structures based on the system's extension status, the following function is provided:
+
+```
+void __init_riscv_feature_bits(void *);
+```
+
+The `__init_riscv_feature_bits` function updates `length`, `mvendorid`, `marchid`, `mimpid` and the `features` in `__riscv_feature_bits` and `__riscv_vendor_feature_bits` according to the enabled extensions in the system.
+
+The `__init_riscv_feature_bits` function accepts an argument of type `void *`. This argument allows the platform to provide pre-computed data and access it without additional effort. For example, Linux could pass the vDSO object to avoid an extra system call.
```

Comment:
Who's expected to call this, and so who's expected to know what magic opaque blob to pass here? Surely you don't expect the compiler to be generating OS-specific code to get data out of the vDSO and pass it to the function?

Reply:
Ping...

Reply:
@BeMg ping